CIP-0147? | Wallet Connector - Own-data wallet (#986)
cardano-foundation/CIPs: CIP-XXXX/README.md

---
CIP: XXX
Title: Own-data Wallet Extension
Category: Tools
Status: Proposed
Authors:
  - Giovanni Garufi <giovanni@mlabs.city>
Implementors: []
Discussions:
    - https://github.com/cardano-foundation/cips/pulls/957
Created: 2024-02-05
License: CC-BY-4.0
Solution-To: CPS-0010
---

## Abstract

In this CIP we build upon the work laid out in [CIP-0144 | Wallet Connector API](https://github.com/cardano-foundation/CIPs/pull/957), which introduces a CIP-30 compatible wallet connection API. Our contribution is to define an extension which, when enabled, would result in what [CPS-0010](https://github.com/cardano-foundation/CIPs/blob/master/CPS-0010/README.md) defines as an "own-data wallet".

## Motivation: why is this CIP necessary?

> [CIP-0144 | Wallet Connector API](link) introduces a new wallet connection API. However, a wallet can not do much without enabling some extensions. Here we present an extension that is equivalent to the familiar CIP-30 (In particular it implements the [full-api](https://cips.cardano.org/cip/CIP-30#full-api) section of CIP-30). Enabling this extension would give users a wallet that has the same capabilities as a CIP-30 wallet, this is defined as an "own-data wallet" in CPS-10.

Author: ```suggestion
[CIP-0144 | Wallet Connector API](https://github.com/cardano-foundation/CIPs/pull/957) introduces a new wallet connection API. However, a wallet can not do much without enabling some extensions. Here we present an extension that is equivalent to the familiar CIP-30 (In particular it implements the [full-api](https://cips.cardano.org/cip/CIP-30#full-api) section of CIP-30). Enabling this extension would give users a wallet that has the same capabilities as a CIP-30 wallet, this is defined as an "own-data wallet" in CPS-10.
```

> [CIP-0144 | Wallet Connector API](https://github.com/cardano-foundation/CIPs/pull/957) introduces a new wallet connection API. However, a wallet can not do much without enabling some extensions. Here we present an extension that is equivalent to the familiar CIP-30 (In particular it implements the [full-api](https://cips.cardano.org/cip/CIP-30#full-api) section of CIP-30). Enabling this extension would give users a wallet that has the same capabilities as a CIP-30 wallet, this is defined as an "own-data wallet" in CPS-10.

Reviewer: ```suggestion
[CIP-0144? | Wallet Connector API](https://github.com/cardano-foundation/CIPs/pull/957) introduces a new wallet connection API. However, a wallet can not do much without enabling some extensions. Here we present an extension that is equivalent to the familiar CIP-30 (In particular it implements the [full-api](https://cips.cardano.org/cip/CIP-30#full-api) section of CIP-30). Enabling this extension would give users a wallet that has the same capabilities as a CIP-30 wallet, this is defined as an "own-data wallet" in CPS-10.
```
The `?` character will make it easier for us to flag if & when the CIP is merged... and is less obtrusive than using the word "Candidate" everywhere.  In the meantime it's necessary to call the reader's or implementor's attention to the fact that this document is not a CIP yet.

## Specification

> Following what is outlined in CIP-0144 we will use the same transport-agnostic format to define the API. See [here](link/#Transport-agnostic-operations) for more details.

Author: ```suggestion
Following what is outlined in CIP-0144 we will use the same transport-agnostic format to define the API. See [here](https://github.com/nazrhom/CIPs/blob/nazrhom/full-data-wallet-connector/CIP-0144/README.md#transport-agnostic-operations) for more details.
```

- `ProofGeneration`: (1) Wallet could not sign the data (e.g. does not have the secret key associated with the address).
- `AddressNotPK`: (2) Address was not a P2PK address and thus had no SK associated with it.
- `UserDeclined`: (3) User declined to sign the data.

##### TxSendError

```
{
  "type": "object",
  "title": "TxSendError",
  "properties": {
    "code": {
      "type": "number",
      "title": "TxSendErrorCode"
      "enum": [1, 2]
    },
    "info": {
      "type": "string"
    }
  },
  "required": ["code"],
}
```

- `Refused`: (1) Wallet refuses to send the tx (could be rate limiting).
- `Failure`: (2) Wallet could not send the tx.

##### TxSignError

```
{
  "type": "object",
  "title": "TxSignError",
  "properties": {
    "code": {
      "type": "number",
      "title": "TxSignErrorCode"
      "enum": [1, 2]
    },
    "info": {
      "type": "string"
    }
  },
  "required": ["code"],
}
```

- `ProofGeneration`: (1) User has accepted the transaction sign, but the wallet was unable to sign the transaction (e.g. not having some of the private keys).
- `UserDeclined`: (2) User declined to sign the transaction.

## Rationale: how does this CIP achieve its goals?

> The goal of this CIP is to define an extension to enable an "own-data wallet". We build upon the work done in [CIP-0144 | Wallet Connector API](link) and define an extension compatible with that CIP.

Author: ```suggestion
The goal of this CIP is to define an extension to enable an "own-data wallet". We build upon the work done in [CIP-0144 | Wallet Connector API](https://github.com/cardano-foundation/CIPs/pull/957) and define an extension compatible with that CIP.
```

## Path to Active

> CIP-0144 includes a [section](link/#CIP-30-Backwards-compatibility) which introduces the idea of a wrapper, that takes a CIP-30 compatible API, and makes it CIP-0144 compatible. Since in this CIP we are simply replicating the functionality offered by CIP-30, something very similar would be possible here. Implementors of this wrapper must take care of the following differences between this CIP and CIP-30.

Author: ```suggestion
CIP-0144 includes a [section](https://github.com/nazrhom/CIPs/blob/nazrhom/full-data-wallet-connector/CIP-0144/README.md#cip-30-backwards-compatibility) which introduces the idea of a wrapper, that takes a CIP-30 compatible API, and makes it CIP-0144 compatible. Since in this CIP we are simply replicating the functionality offered by CIP-30, something very similar would be possible here. Implementors of this wrapper must take care of the following differences between this CIP and CIP-30.
```

> ## Path to Active
> 
> CIP-0144 includes a [section](https://github.com/nazrhom/CIPs/blob/nazrhom/full-data-wallet-connector/CIP-0144/README.md#cip-30-backwards-compatibility) which introduces the idea of a wrapper, that takes a CIP-30 compatible API, and makes it CIP-0144 compatible. Since in this CIP we are simply replicating the functionality offered by CIP-30, something very similar would be possible here. Implementors of this wrapper must take care of the following differences between this CIP and CIP-30.
> 
> - JSON is used instead of CBOR. The wrapper must take care of translating arguments and results.
> - This API has no pagination, while CIP-30 generally does. The wrapper can either implement pagination locally, or error and inform the user if they request it.
> - In the original CIP-30, the `api` object returned has, on the top-level, the methods that we expect on `api.cip30`. The wrapper must take care to translate calls appropriately.
> - CIP-30 endpoints will return `null` in some situations where the request is correct, but not satisfiable. In this CIP we introduce an error code specifically for that. The wrapper will need to check some results for `null` and throw an appropriate error.
> 

Reviewer: ```suggestion
### Backward Compatibility

CIP-0144 includes a [section](https://github.com/nazrhom/CIPs/blob/nazrhom/full-data-wallet-connector/CIP-0144/README.md#cip-30-backwards-compatibility) which introduces the idea of a wrapper, that takes a CIP-30 compatible API, and makes it CIP-0144 compatible. Since in this CIP we are simply replicating the functionality offered by CIP-30, something very similar would be possible here. Implementors of this wrapper must take care of the following differences between this CIP and CIP-30.

- JSON is used instead of CBOR. The wrapper must take care of translating arguments and results.
- This API has no pagination, while CIP-30 generally does. The wrapper can either implement pagination locally, or error and inform the user if they request it.
- In the original CIP-30, the `api` object returned has, on the top-level, the methods that we expect on `api.cip30`. The wrapper must take care to translate calls appropriately.
- CIP-30 endpoints will return `null` in some situations where the request is correct, but not satisfiable. In this CIP we introduce an error code specifically for that. The wrapper will need to check some results for `null` and throw an appropriate error.

## Path to Active

```
This section is in the correct order but really part of the _Rationale_ since the _Path to Active_ is a series of checklists after you have outlined any dependencies in your chosen method... as this section does very well.  `Backward Compatibility` was part of the original CIP format indicated in CIP-0001 until it became obsolete as generally being part of the _Rationale_... so generally it's a subsection there.

---

> CIP: XXX

Reviewer: ```suggestion
CIP: 147
```

Reviewer: resolved with https://github.com/cardano-foundation/CIPs/pull/986/commits/58875ec56a9da37c8c0e092a905be99bd0993e72

> Title: Own-data Wallet Extension

Reviewer: ```suggestion
Title: Wallet Connector - Own-data wallet
```

> Category: Tools

Reviewer: ```suggestion
Category: Wallets
```
@Ryun1 although you confirmed that the resulting processes will still be "wallets" even though functionally separate from web interfaces, wouldn't therefore the category `Wallets` still be appropriate here rather than the perhaps too general `Tools`?

Reviewer: resolved with https://github.com/cardano-foundation/CIPs/pull/986/commits/58875ec56a9da37c8c0e092a905be99bd0993e72

Status: Proposed
Authors:
  - Giovanni Garufi <giovanni@mlabs.city>
Implementors: []
Discussions:

>     - https://github.com/cardano-foundation/cips/pulls/957

Reviewer: ```suggestion
    - https://github.com/cardano-foundation/CIPs/pull/986
```
@nazrhom I'm guessing this may have come from copying CIP-0144, and that it just needs to be updated to (this) current PR number?  I don't think it's necessary to include the CIP-0144 discussion with this PR since people will naturally trace the origin of the CIP from links here in this document — and, more significantly, the taxonomy in the CIP title.

Reviewer: resolved with https://github.com/cardano-foundation/CIPs/pull/986/commits/58875ec56a9da37c8c0e092a905be99bd0993e72

Created: 2024-02-05
License: CC-BY-4.0

> Solution-To: CPS-0010

Reviewer: ```suggestion
```
Although candidate CIP-0144 _is_ definitely a solution to CPS-0010, I would argue that this CIP is not... because it's an extension to an API rather than an API itself.  Again, this may have been inadvertently copied... but if not, I would argue that including links to indirectly related CPSs could make the front matter detail generally longer and more difficult to maintain and navigate.

Reviewer: resolved with https://github.com/cardano-foundation/CIPs/pull/986/commits/58875ec56a9da37c8c0e092a905be99bd0993e72

- [appendix](#appendix) in which we define schemas for types required by this extension and the error types

We will use an identifier in the anchor to refer to the schema where each type is defined. For example, if we want to reference the `Transaction` type, as defined in CIP-116 we will use the following schema reference `{ "$ref": "#/cip-116/Transaction" }`.

If an operation does not require any arguments as part of it's request, or does not return a meaningful response, we will represent this as an `{}` on the respective operation field. The arguments to an operation will always be represented as an object. If a field is not marked as `required`, then that argument is to be considered optional.

For each operation we will provide some details on how the implementation should behave. Note that some of these are taken verbatim from CIP-30.

> #### CIP-XXXX

Reviewer: ```suggestion
### CIP-XXXX
```
Since we seemed to have skipped the H3 indentation level, I think the headings below also need to be increased 1 level (until the subsection on "additional schemas").

Reviewer: resolved with https://github.com/cardano-foundation/CIPs/pull/986/commits/a6fb968764c61125441318388f95e8eb3d109ea7

  ]
}
```

As wallets should already have this ability, we allow dApps to request that a transaction be sent through it. If the wallet accepts the transaction and tries to send it, it shall return the transaction id for the dApp to track. The wallet is free to return the `TxSendError` with error code `Refused` if they do not wish to send it, or `Failure` if there was an error in sending it (e.g. preliminary checks failed on signatures).

##### ApiVersion

```
{
  "operation": "apiVersion",
  "request": {},
  "response": { "$ref": "#/appendix/SemVer" },
  "errors": []
}
```

Returns the API version for the CIP-XXXX extension. This must correspond to the value of `CIP-XXXX` specified in this document, appropriately transformed into a `SemVer` object.

### Versioning

The version of this extension will be tracked in the table below.

While the CIP is in preparation, the version shall be set to `0.0.0`. The moment this CIP is merged the version shall be set to `1.0.0`, and all implementations should consider that the current version. Any changes to the connection API should come in form of PRs to this CIP.

| API | Version |
| --- | --- |
| CIP-XXXX | 0.0.0 |

> ### Appendix
> 
> This appendix contains additional schemas for types that are used in the APIs.

Reviewer: ```suggestion
### Additional schemas
```
Calling this an _Appendix_ would make many readers think this is the end of the document, rather than a appendical part of the _Specification_ which I believe you intended & which makes more sense.

Reviewer: resolved with https://github.com/cardano-foundation/CIPs/pull/986/commits/58875ec56a9da37c8c0e092a905be99bd0993e72
